File: 2d-dp/518-Coin-Change-II/neet.js
var change = function (amount, coins) {
  const cache = {};
  const getKey = (i, a) => `${i}_${a}`;

  const dfs = (i, a) => {
    const key = getKey(i, a);
    if (a === amount) return 1;
    if (a > amount || i === coins.length) return 0;
    if (cache[key] !== undefined) return cache[key];

    cache[key] = dfs(i, a + coins[i]) + dfs(i + 1, a);
    return cache[key];
  };

  return dfs(0, 0);
};

File: 2d-dp/518-Coin-Change-II/neet.py
class Solution:
    def change(self, amount: int, coins) -> int:
        cache = {}

        def dfs(i, a):
            if a == amount: return 1
            if a > amount or i == len(coins): return 0
            if (i, a) in cache: return cache[(i, a)]

            cache[(i, a)] = dfs(i, a + coins[i]) + dfs(i + 1, a)
            return cache[(i, a)]

        return dfs(0, 0)